Moved the links to the sidebar #236

LeoLuosifen opened the pull request:
## Summary

This PR refactors the layout by moving the navigation links previously displayed in the top banner into a collapsible sidebar (`dmc.Drawer`) on the left side of the page. It also updates the end-to-end test file `spec.cy.js` to reflect the new layout and interaction behavior.

## Changes

- Replaced top tab navigation with a collapsible sidebar.
- Used `dash-mantine-components` (`dmc.NavLink`, `dmc.SegmentedControl`, etc.) to build the new sidebar.
- Moved tab navigation links (e.g., *Select Weather File*, *Climate Summary*, etc.) into the sidebar under “Pages Menu”.
- Moved **SI/IP** and **Global/Local Value Ranges** controls into the sidebar under “Tools Menu”.
- Ensured the sidebar can be toggled via a burger button and closes on navigation.
- Simplified the top banner layout (title, logo, location only).
- **Updated `spec.cy.js` test code** to match the new layout structure:
  - Adjusted selectors and actions to open the sidebar before interacting with nav links or controls.
  - Ensured UI controls are still functional in their new locations.

LeoLuosifen committed:
refactor(layout): move tabs links to collapsible sidebar using dash-mantine-components
- updated dash and dash-mantine-components
- moved the navigation and document, global, and IP to the sidebar
- modified the spec.cy.js
- updated the tabs.css and layout.css

LeoLuosifen committed:
style: formatted the code via using ruff and black.

LeoLuosifen committed:
fix: fixed the comment

LeoLuosifen committed:
fix: re-formatted the file via using ruff

FedericoTartarini commented on assets/tabs.css:
```diff
@@ -285,4 +285,134 @@ p {
+}
+
+/* side bar */
+#sidebar {
```
please change this. We should not be using all this CSS, just use it if stricly necessary. This is not necessary and will need to be removed

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -136,13 +137,13 @@ def section_one():
                                 n_clicks=0,
                             ),
-                            html.Div(
+                            dmc.Box(
```
here most likely we want to remove this div all together not just use a dmc.Box instead. We should simplify the code as much as possible and remove the classes. We should only be using mantine and apply styles like padding and margins, if needed in line

FedericoTartarini commented on pages/lib/layout.py:
```diff
@@ -30,52 +41,51 @@ def alert():
-                id=ElementIds.ID_LAYOUT_INTERVAL_COMPONENT,
-                interval=12 * 1000,
-                n_intervals=0,
+            dmc.Box(
```
why did you do this? Internal I believe it is needed to refresh the page. If not needed remove it but do not replace it with a box. Please be very careful when making changes otherwise the code will break

FedericoTartarini commented on pages/lib/layout.py:
```diff
@@ -85,145 +95,221 @@ def footer():
+
+
+# Pages Icon
+PAGE_ICON_MAP = {
```
use classess never a dictionaries

FedericoTartarini commented on pages/natural_ventilation.py:
```diff
@@ -74,13 +75,13 @@ def update_layout(si_ip):
             type="circle",
         ),
-        html.Div(
+        dmc.Box(
```
same as previous comment, I do not think we need all these box. I never use them in my code

FedericoTartarini reviewed (changes requested):
I left a few comments, I did not review the whole code. In summary do not use CSS unless 100% needed, it is generally not needed since mantine components are well designed. Also do not simply replace Div with Box, but instead check if the Div is needed at all, if not just remove it. I am sure that more than 50% of them can be removed by using the right mantine components like Stack, Group, Grid, ...

FedericoTartarini commented on assets/layout.css:
```diff
@@ -29,4 +29,15 @@
 
+/* begin displaying banner below the banner */
+.custom-sidebar .mantine-Drawer-content {
+    top: 80px !important;
```
1rem

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -92,41 +93,41 @@ def section_one():
             ),
-            html.Div(
+            dmc.Box(
                 className="container-row justify-content-center",
```
here example where we may not need it

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -168,11 +169,11 @@ def section_one():
                             ),
-                            html.Div(
+                            dmc.Box(
                                 className="container-row justify-center",
```
check if you can replace with SimpleGrid

FedericoTartarini commented on pages/psy-chart.py:
```diff
@@ -61,13 +61,13 @@
     """"""
-    return html.Div(
+    return dmc.Box(
         className="container-row full-width three-inputs-container",
```
dmc.Grid(
    children=[
        dmc.GridCol(dmc.Box("1", style=style), span=6),
        dmc.GridCol(dmc.Box("2", style=style), span=3),
        dmc.GridCol(dmc.Box("3", style=style), span=3),
    ],
    gutter="xl",
)

FedericoTartarini replied:
remove the class name

FedericoTartarini commented on pages/psy-chart.py:
```diff
@@ -61,13 +61,13 @@
         className="container-row full-width three-inputs-container",
         children=[
-            html.Div(
+            dmc.Box(
```
dmc.GridCol(

FedericoTartarini commented on pages/psy-chart.py:
```diff
@@ -61,13 +61,13 @@
         children=[
-            html.Div(
+            dmc.Box(
                 className=container_col_center_one_of_three,
```
remove the class name

FedericoTartarini commented on assets/tabs.css:
```diff
@@ -285,4 +285,134 @@ p {
+    position: fixed;
+    top: 25px;
+    right: 10px;
+    width: 400px;
```
https://www.dash-mantine-components.com/style-props

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -64,7 +65,7 @@
     """Return the inputs from section one."""
-    return html.Div(
+    return dmc.Box(
         className="container-row full-width row-center",
```
check how to remove the class.

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -64,7 +65,7 @@
+    return dmc.Box(
         className="container-row full-width row-center",
         children=[
             html.H4(className="text-next-to-input", children=["Select a variable: "]),
```
dmc.Title(order 4 ) no css

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -168,11 +169,11 @@ def section_one():
+                            dmc.Box(
                                 className="container-row justify-center",
                                 children=[
                                     html.H6("Hour Range", style={"flex": "20%"}),
```
same comment

LeoLuosifen committed:
refactor: remove unnecessary CSS/HTML
- used Dash-Mantine components styles API
- formatted the code via using ruff and black

LeoLuosifen committed:
add: created a new class file to store icon value

LeoLuosifen committed:
fix: removed all html and dash bootstrap via using dash mantine components

LeoLuosifen committed:
fix: fixed the components that ensured cypress testing could be passed

Sallie00 commented:
After thorough investigation, we have traced the issue back to the `utils.py` file.

We discovered that replacing the `html.A` component with its Dash Mantine equivalent (e.g., dmc.Anchor) causes unexpected redirects to the home page, especially during Cypress tests. This prevents the `/summary` page from loading properly and makes it impossible to verify elements on that page.

To ensure stability, we have kept this single use of `html.A` unchanged. All other HTML elements have been successfully migrated to Dash Mantine Components.

Please let us know how you’d prefer to proceed with this specific case. 
@FedericoTartarini

FedericoTartarini commented:
@Sallie00 I am happy with your decision. please keep those html.A components when necessary and otherwise breaking the code. I like this since we can still use the old test framework. Please tag me here once the pull request is complete so I will review it.

Sallie00 commented:
Hi @FedericoTartarini , The pull request is now ready for your review.

FedericoTartarini commented:
I really like the work; however, I noticed a major issue. The navbar is closed by default, and it will be very confusing for current and future users to find the tabs. The navbar, to work well:
1. needs to be open by default
2. it should not be going over the page but shift the page as I mentioned before, see this website as an example https://fly-log.netlify.app/
3. this is very important also because it is a bit annoying to have to open the navbar each time the user wants to change the page.
4. the logo in the title of the navbar is not showing
5. reduce sligthly the spacing between the links in the navbar.
Please implement these changes and then I will carefully review the pull request.

FedericoTartarini commented on pages/lib/page_icon.py:
```diff
@@ -0,0 +1,32 @@
+class PageIcon:
```
please use a better name for this variable, like NavBarIcons. Moreover, i am not sure we need a new file for this

FedericoTartarini commented on pages/lib/layout.py:
```diff
@@ -43,187 +59,330 @@ def alert():
                 ],
-                width=12,
-                md=8,
+                flex="0 0 66.666667%",
```
this is not needed I am sure, if it is we need to find a better code to do the same, like simpleGrid or Grid. Please fix

FedericoTartarini commented on pages/lib/layout.py:
```diff
@@ -43,187 +59,330 @@ def alert():
-                md=8,
+                flex="0 0 66.666667%",
+                maw="66.666667%",
+                p="0px 15px 10px 15px",
```
please simplify

FedericoTartarini commented on pages/lib/layout.py:
```diff
@@ -43,187 +59,330 @@ def alert():
+        zIndex=1001,
+        opened=False,
+        styles={
+            "content": {
```
too much style, please remove

FedericoTartarini commented on pages/lib/layout.py:
```diff
@@ -236,48 +395,79 @@ def store():
+)
+def show_alert_after_delay(n_intervals):
+    """Show alert after 6 seconds, then hide after 5 more seconds"""
+    base_style = {
```
please remove all these styles, use default style of the components

FedericoTartarini commented on pages/lib/utils.py:
```diff
@@ -147,38 +147,40 @@ def generate_custom_inputs_psy(
+            align="center",
+            mt="md",
+            px="md",
+            children=[
```
can't we simply just return the title and not wrap it in a group?

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -64,10 +62,14 @@
-    return html.Div(
-        className="container-row full-width row-center",
+    return dmc.Group(
+        align="center",
```
very good that you removed the class. Have you deleted the class from the .css file?

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -79,302 +81,367 @@ def section_one_inputs():
+                tooltip_text="count, mean, std, min, max, and percentiles",
+                id_button=IdButtons.TABLE_EXPLORE,
+            ),
+            dmc.Center(
```
do we need the center or can it be set using align in the Stack?

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -79,302 +81,367 @@ def section_one_inputs():
+                        ),
+                        dmc.Grid(
+                            gutter="sm",
+                            align="center",
```
there are many align center, do we need them all?

FedericoTartarini commented on pages/explorer.py:
```diff
@@ -79,302 +81,367 @@ def section_one_inputs():
+                                    ],
+                                ),
+                                dmc.Grid(
+                                    gutter="sm",
```
same comment as below. Do we really need all these gutters and align?

FedericoTartarini replied:
can't we just use the defaults?

FedericoTartarini commented on pages/natural_ventilation.py:
```diff
@@ -41,13 +39,7 @@
-            #
-        ],
-    )
+    return dmc.Stack(id=ElementIds.MAIN_NV_SECTION, gap="md")
```
do we need the gap here?

FedericoTartarini commented on pages/outdoor.py:
```diff
@@ -36,184 +36,217 @@
 def inputs_outdoor_comfort():
-    return dbc.Row(
-        className="container-row full-width three-inputs-container",
+    return dmc.Grid(
```
do we really need all these nested Grid -> GridCol -> Grid -> GridCol. I do not think so, please carefully review the old bad code and improve it, where possible. i only left this comment but please review all the files.

FedericoTartarini commented on pages/psy-chart.py:
```diff
@@ -61,154 +58,203 @@
+                                    gutter="sm",
+                                    align="center",
+                                    children=[
+                                        dmc.GridCol(
```
same as previous comment about all these possibly unecessary levels of nesting

FedericoTartarini commented on pages/select.py:
```diff
@@ -50,62 +48,85 @@ def layout():
+        mx=0,
+        px=0,
+        py="md",
+        style={
```
please try to remove

ziqi0067 replied:
@FedericoTartarini  The optimal approach here is to remove "mx", "px", and "py", but the style must be retained. Otherwise, the page layout will drastically change.

FedericoTartarini commented on pages/summary.py:
```diff
@@ -231,94 +200,76 @@ def update_location_info(ts, df, meta, si_ip):
-            ),
-            dbc.Row(total_diffuse_rad),
+    return dmc.Stack(
+        gap=4,
```
do not use numbers ever, use only the strings "md", "sm", ...

FedericoTartarini commented on pages/summary.py:
```diff
@@ -231,94 +200,76 @@ def update_location_info(ts, df, meta, si_ip):
     [
         Output(ElementIds.DEGREE_DAYS_CHART_WRAPPER, "children"),
-        Output(ElementIds.WARNING_CDD_HIGHER_HDD, "is_open"),
+        Output(ElementIds.WARNING_CDD_HIGHER_HDD, "children"),
```
be very very carefull, why are you changing these?

FedericoTartarini commented on pages/t_rh.py:
```diff
@@ -30,69 +32,66 @@
-            html.Div(
-                className="container-col",
+            dmc.Stack(
+                gap="lg",
```
please keep consistent gaps, sometimes you use small sometimes large. I would just use the defaults, so we can have less code and keep consistency in the layout

FedericoTartarini commented on pages/t_rh.py:
```diff
@@ -30,69 +32,66 @@
                     dcc.Loading(
                         type="circle",
-                        children=html.Div(id=ElementIds.DAILY),
+                        children=dmc.Stack(id=ElementIds.DAILY, gap=0),
```
why this?

FedericoTartarini commented on main.py:
```diff
@@ -1,22 +1,22 @@
-    fluid=True,
-    style={"padding": "0"},
+app.layout = dmc.MantineProvider(
+    theme={"colorScheme": "light", "primaryColor": "blue"},
```
do we need this? Is not this the default?

FedericoTartarini reviewed (changes requested):
@Sallie00 I really like the work; however, I noticed a major issue. The navbar is closed by default, and it will be very confusing for current and future users to find the tabs. The navbar, to work well:
1. needs to be open by default
2. it should not be going over the page but shift the page as I mentioned before, see this website as an example https://fly-log.netlify.app/
3. this is very important also because it is a bit annoying to have to open the navbar each time the user wants to change the page.
4. the logo in the title of the navbar is not showing
5. reduce sligthly the spacing between the links in the navbar.

## code review
Very good work I'm very happy with the progress. However, I think they code can be further simplified. For example, I think you have too many nested components. Please review all my comments carefully. I am a little bit worried because sometimes you change the argument of the input or of the output in the callback. Please make sure that this is what you want to do. Please also try to use as much as possible the default component as provided by Mantine. Using the default will ensure that the website is consistent and beautiful.

LeoLuosifen commented:
@FedericoTartarini
Hi Federico, I've restructured two versions of the layout of the Clima. Both are used with AppShell to manage.

## Version 1
![version_1_1](https://github.com/user-attachments/assets/3641a790-a6c6-4afb-84db-46a01b88a6d0)
![version_1_2](https://github.com/user-attachments/assets/f45a22cf-da01-4362-80c5-424f6b78af02)

The first version uses the Full AppShell Layout https://py.cafe/dash.mantine.components/AppShell-with-all-elements, divided into header, navbar, main, and footer sections. The main section is scrollable, though this may not be readily apparent in my screenshot. Simultaneously, clicking the burger button controls the display and concealment of the sidebar, functioning identically to the demonstration in Dash Mantine. By default, it remains expanded.

## Version 2
![version_2_1](https://github.com/user-attachments/assets/29eed3e1-103a-4f6d-8003-f838e457e652)
![version_2_2](https://github.com/user-attachments/assets/15b986d4-a9f5-4b31-8386-c061ac0536ac)

The secondary version uses Collapsible AppShell Layout  https://py.cafe/dash.mantine.components/Collapsible-navbar-on-both-desktop-and-moble, divided into header, navbar, and main sections. The main section includes the original layout's main and footer elements and is also scrollable. The function of the burger button remains unchanged from Version 1.

Which version do you prefer?

FedericoTartarini commented:
@LeoLuosifen I am not fully understanding what the difference is between v1 and v2. Please explain the key differences and how this will affect the user experience of Clima.

LeoLuosifen commented:
@FedericoTartarini The entire page appears somewhat small in v1, as my computer screen is rather compact. While it might look better on a larger display, this may not suit all users. The v2 more closely looks like the effect shown in the link you provided  https://fly-log.netlify.app/.

FedericoTartarini commented:
I want to recreate the same effect as in https://fly-log.netlify.app/. please also note that the navbar width can be slighlty reduced since it does not need to be so wide. If version 2 is similar to https://fly-log.netlify.app/ then we can implement version 2

yluo0664 committed:
refactor: used AppShell in Dash Mantine to manage the whole layout
- optimized the original simple dmc components
- removed the element ids are no longer needed
- removed the unnecessary css style in css files
- removed page_icon.py, moved to layout.py and renamed class name to NavbarIcons
- updated spec.cy.js
- reformatted the code

yluo0664 committed:
fix: removed the styles in the function show_alert_after_delay while using default style

yluo0664 committed:
fix: removed the unnecessary style and optimized spacing

LeoLuosifen commented:
@FedericoTartarini 
Hi Federico, we've removed the unnecessary spacing and made the default sidebar of the pages menu open, also ensured the sidebar width is consistent. Could you kindly review it and let us know if there's anything else that could be improved?
![sidebar_1](https://github.com/user-attachments/assets/0d5a352f-02c9-4c22-b50b-507a732bd012)
![sidebar_2](https://github.com/user-attachments/assets/b88ab93f-531a-4bce-993c-a8c9393cd44a)

FedericoTartarini committed:
style(layout): Simplify children structure in ScrollArea

Removed unnecessary Stack component to streamline the layout
and improve readability. Also optimized spacing in the layout
by removing redundant CSS properties.

FedericoTartarini committed:
style(select): Refactor layout to use Stack component

Replace Box with Stack for improved layout structure and
remove unnecessary style properties to simplify the code.

FedericoTartarini committed:
refactor(layout): Simplify layout structure in summary and template graphs

Consolidate container and stack components for improved readability and
maintainability. Adjust grid column spans for responsive design.

FedericoTartarini committed:
style(utils): Remove unnecessary margin and padding in tooltip functions

Clean up the layout by eliminating redundant margin and padding properties
in the title_with_tooltip function to streamline the component's styling.

FedericoTartarini committed:
refactor(layout): Replace Container with Stack for layout consistency

Updated the layout function to utilize the Stack component instead of
Container, improving the structure and alignment of child elements.

FedericoTartarini committed:
refactor(layout): Move layout function to improve structure

Refactor the layout function for tab four by removing the old definition
and ensuring the layout is consistently defined using the Stack component.

FedericoTartarini committed:
refactor(layout): Update loading component structure

Replace dcc.Loading with dmc.Center for improved layout consistency.
Remove unused CSS styles related to tabs for cleaner codebase.

FedericoTartarini committed:
refactor(layout): Replace Box with Stack for tab six layout

Refactor the layout function to use a Stack component instead of Box,
improving the structure and consistency of the layout in tab six.

FedericoTartarini committed:
feat(css): remove tabs.css

FedericoTartarini committed:
refactor(viz): Remove y-axis range setting hhd and ccd

Eliminate fixed y-axis range to allow for dynamic scaling based on data.

FedericoTartarini committed:
refactor(ui): Remove inline styles from dropdown components

Clean up the code by removing hardcoded width styles from various dropdowns
to improve layout consistency and maintainability.

FedericoTartarini committed:
refactor(explorer): Clean up section one layout code

Reorganize the layout code in section one for better readability and maintainability.
This includes consistent indentation and formatting adjustments.

FedericoTartarini commented:
Very good work, I really like what you have done. 

Please all carefully review all my changes. This will be an opportunity for you to learn and see how css could have been removed and the code could have been much further simplified. I did not implemented all the changes I wanted but as you can see we could much furhter improve the code and avoiding nesting components. I will need to stop here for today and I will accept the pull request for now, but note that we can improve it much further.
